File: pkg/models/query.go
package models

// QueryType defines the query operation
// +enum
type QueryType string

const (
	// QueryTypeCommits is sent by the frontend when querying commits in a GitHub repository
	QueryTypeCommits QueryType = "Commits"
	// QueryTypeIssues is used when querying issues in a GitHub repository
	QueryTypeIssues QueryType = "Issues"
	// QueryTypeContributors is used when querying contributors in a GitHub repository
	QueryTypeContributors QueryType = "Contributors"
	// QueryTypeTags is used when querying tags in a GitHub repository
	QueryTypeTags QueryType = "Tags"
	// QueryTypeReleases is used when querying releases in a GitHub repository
	QueryTypeReleases QueryType = "Releases"
	// QueryTypePullRequests is used when querying pull requests in a GitHub repository
	QueryTypePullRequests QueryType = "Pull_Requests"
	// QueryTypePullRequestReviews is used when querying pull request reviews in a GitHub repository
	QueryTypePullRequestReviews = "Pull_Request_Reviews"
	// QueryTypeLabels is used when querying labels in a GitHub repository
	QueryTypeLabels QueryType = "Labels"
	// QueryTypeRepositories is used when querying for a GitHub repository
	QueryTypeRepositories QueryType = "Repositories"
	// QueryTypeOrganizations is used when querying for GitHub organizations
	QueryTypeOrganizations QueryType = "Organizations"
	// QueryTypeGraphQL is used when sending an ad-hoc graphql query
	QueryTypeGraphQL QueryType = "GraphQL"
	// QueryTypePackages is used when querying for NPM / Docker / etc packages
	QueryTypePackages QueryType = "Packages"
	// QueryTypeMilestones is used when querying for milestones in a repository
	QueryTypeMilestones QueryType = "Milestones"
	// QueryTypeVulnerabilities is used when querying a vulnerability for a repository
	QueryTypeVulnerabilities QueryType = "Vulnerabilities"
	// QueryTypeProjects is used when querying projects for an organization
	QueryTypeProjects QueryType = "Projects"
	// QueryTypeProjectItems is used when querying projects for an organization
	QueryTypeProjectItems QueryType = "ProjectItems"
	// QueryTypeStargazers is used when querying stargazers for a repository
	QueryTypeStargazers QueryType = "Stargazers"
	// QueryTypeWorkflows is used when querying workflows for an organization
	QueryTypeWorkflows QueryType = "Workflows"
	// QueryTypeWorkflowUsage is used when querying a specific workflow usage
	QueryTypeWorkflowUsage QueryType = "Workflow_Usage"
	// QueryTypeWorkflowRuns is used when querying workflow runs for a repository
	QueryTypeWorkflowRuns QueryType = "Workflow_Runs"
	// QueryTypeCodeScanning is used when querying code scanning alerts for a repository
	QueryTypeCodeScanning QueryType = "Code_Scanning"
	// QueryTypeDeployments is used when querying deployments for a repository
	QueryTypeDeployments QueryType = "Deployments"
	// QueryTypeCommitFiles is used when querying files changed in a specific commit
	QueryTypeCommitFiles QueryType = "Commit_Files"
	// QueryTypePullRequestFiles is used when querying files changed in a specific pull request
	QueryTypePullRequestFiles QueryType = "Pull_Request_Files"
)

// Query refers to the structure of a query built using the QueryEditor.
// Every query uses this query type and has to include options for each type of query.
// For example, listing commits can be filtered by author, but filtering contributors by author
// doesn't provide much value, but is included in the query schema anyways.
type Query struct {
	Repository string `json:"repository"`
	Owner      string `json:"owner"`
}

// PullRequestsQuery is used when querying for GitHub Pull Requests
type PullRequestsQuery struct {
	Query
	Options ListPullRequestsOptions `json:"options"`
}

// PullRequestReviewsQuery is used when querying for GitHub Pull Request Reviews
type PullRequestReviewsQuery struct {
	Query
	Options ListPullRequestsOptions `json:"options"`
}

// CommitsQuery is used when querying for GitHub commits
type CommitsQuery struct {
	Query
	Options ListCommitsOptions `json:"options"`
}

// TagsQuery is used when querying for GitHub tags
type TagsQuery struct {
	Query
	Options ListTagsOptions `json:"options"`
}

// LabelsQuery is used when querying for GitHub issue labels
type LabelsQuery struct {
	Query
	Options ListLabelsOptions `json:"options"`
}

// ReleasesQuery is used when querying for GitHub issue labels
type ReleasesQuery struct {
	Query
	Options ListReleasesOptions `json:"options"`
}

// ContributorsQuery is used when querying for GitHub contributors
type ContributorsQuery struct {
	Query
	Options ListContributorsOptions `json:"options"`
}

// RepositoriesQuery is used when querying for GitHub repositories
type RepositoriesQuery struct {
	Query
}

// IssuesQuery is used when querying for GitHub issues
type IssuesQuery struct {
	Query
	Options ListIssuesOptions `json:"options"`
}

// PackagesQuery is used when querying for GitHub packages, including NPM, Maven, PyPi, Rubygems, and Docker
type PackagesQuery struct {
	Query
	Options ListPackagesOptions `json:"options"`
}

// MilestonesQuery is used when querying for GitHub milestones
type MilestonesQuery struct {
	Query
	Options ListMilestonesOptions `json:"options"`
}

// VulnerabilityQuery is used when querying for GitHub Repository Vulnerabilities
type VulnerabilityQuery struct {
	Query
	Options ListVulnerabilitiesOptions `json:"options"`
}

// StargazersQuery is used when querying stargazers for a repository
type StargazersQuery struct {
	Query
}

// WorkflowsQuery is used when querying workflows for an organization
type WorkflowsQuery struct {
	Query
	Options ListWorkflowsOptions `json:"options"`
}

// WorkflowUsageQuery is used when querying a workflow usage
type WorkflowUsageQuery struct {
	Query
	Options WorkflowUsageOptions `json:"options"`
}

// WorkflowRunsQuery is used when querying workflow runs for a repository
type WorkflowRunsQuery struct {
	Query
	Options WorkflowRunsOptions `json:"options"`
}

// CodeScanningQuery is used when querying code scanning alerts for a repository
type CodeScanningQuery struct {
	Query
	Options CodeScanningOptions `json:"options"`
}

// DeploymentsQuery is used when querying deployments for a repository
type DeploymentsQuery struct {
	Query
	Options ListDeploymentsOptions `json:"options"`
}

// OrganizationsQuery is used when querying for GitHub organizations
type OrganizationsQuery struct {
}

// CommitFilesQuery is used when querying for files changed in a GitHub commit
type CommitFilesQuery struct {
	Query
	Options CommitFilesOptions `json:"options"`
}

// PullRequestFilesQuery is used when querying for files changed in a GitHub pull request
type PullRequestFilesQuery struct {
	Query
	Options PullRequestFilesOptions `json:"options"`
}

File: src/types/query.ts
import { type DataQuery } from '@grafana/schema';
import { PullRequestTimeField, IssueTimeField, WorkflowsTimeField, PackageType, ProjectQueryType, QueryTypes } from '../constants';
import type { Filter } from 'components/Filters';

export type QueryType = typeof QueryTypes[number]

export type Options = { [index: string]: any; }

export type RepositoryOptions = { repository?: string; owner?: string; }

type BaseQuery<T extends QueryType, O extends Options> = { queryType: T; options?: O; } & RepositoryOptions & DataQuery;

//#region Code_Scanning Query
export type CodeScanningOptions = Options & {
  gitRef?: string;
  state?: string
}
type Code_ScanningQuery = BaseQuery<'Code_Scanning', CodeScanningOptions>
//#endregion

//#region Commits Query
export type CommitsOptions = Options & {
  gitRef?: string;
}
type CommitsQuery = BaseQuery<'Commits', CommitsOptions>
//#endregion

//#region Commit_Files Query
export type CommitFilesOptions = Options & {
  commitSha?: string;
}
type Commit_FilesQuery = BaseQuery<'Commit_Files', CommitFilesOptions>
//#endregion

//#region Issues Query
export type IssuesOptions = Options & {
  timeField?: IssueTimeField;
  query?: string;
}
type IssuesQuery = BaseQuery<'Issues', IssuesOptions>
//#endregion

//#region Contributors Query
export type ContributorsOptions = Options & {
  query?: string;
}
type ContributorsQuery = BaseQuery<'Contributors', ContributorsOptions>
//#endregion

//#region Tags Query
export type TagsOptions = Options & {}
type TagsQuery = BaseQuery<'Tags', TagsOptions>
//#endregion

//#region Releases Query
export type ReleasesOptions = Options & {}
type ReleasesQuery = BaseQuery<'Releases', ReleasesOptions>
//#endregion

//#region Pull_Requests Query
export type Pull_RequestsOptions = Options & {
  timeField?: PullRequestTimeField;
  query?: string;
}
type Pull_RequestsQuery = BaseQuery<'Pull_Requests', Pull_RequestsOptions>
//#endregion

//#region Pull_Request_Files Query
export type Pull_Request_FilesOption = Options & {
  prNumber?: number;
}
type Pull_Request_FilesQuery = BaseQuery<'Pull_Request_Files', Pull_Request_FilesOption>
//#endregion

//#region Pull_Request_Reviews Query
export type PullRequestReviewsOptions = Options & {
  timeField?: PullRequestTimeField;
  query?: string;
}
type Pull_Request_ReviewsQuery = BaseQuery<'Pull_Request_Reviews', PullRequestReviewsOptions>
//#endregion

//#region Labels Query
export type LabelsOptions = Options & {
  query?: string;
}
type LabelsQuery = BaseQuery<'Labels', LabelsOptions>
//#endregion

//#region Repositories Query
type RepositoriesQuery = BaseQuery<'Repositories', {}>
//#endregion

//#region Organizations Query
type OrganizationsQuery = BaseQuery<'Organizations', {}>
//#endregion

//#region GraphQL Query
type GraphQLQuery = BaseQuery<'GraphQL', {}>
//#endregion

//#region Milestones Query
export type MilestonesOptions = Options & {
  query?: string;
}
type MilestonesQuery = BaseQuery<'Milestones', MilestonesOptions>
//#endregion

//#region Packages Query
export type PackagesOptions = Options & {
  names?: string;
  packageType?: PackageType;
}
type PackagesQuery = BaseQuery<'Packages', PackagesOptions>
//#endregion

//#region Vulnerabilities Query
type VulnerabilitiesQuery = BaseQuery<'Vulnerabilities', {}>
//#endregion

//#region Projects Query
export type ProjectsOptions = Options & {
  organization?: string;
  number?: number | string;
  user?: string;
  kind?: ProjectQueryType;
  filters?: Filter[];
}
type ProjectsQuery = BaseQuery<'Projects', ProjectsOptions>
//#endregion

//#region ProjectItems Query
type ProjectItemsQuery = BaseQuery<'ProjectItems', {}>
//#endregion

//#region Stargazers Query
type StargazersQuery = BaseQuery<'Stargazers', {}>
//#endregion

//#region Workflow_Runs Query
export type WorkflowRunsOptions = Options & {
  workflowID?: string;
  branch?: string;
}
type Workflow_RunsQuery = BaseQuery<'Workflow_Runs', WorkflowRunsOptions>
//#endregion

//#region Workflow_Usage Query
export type WorkflowUsageOptions = Options & {
  workflowID?: number;
}
type Workflow_UsageQuery = BaseQuery<'Workflow_Usage', WorkflowUsageOptions>
//#endregion

//#region Workflows Query
export type WorkflowsOptions = Options & {
  timeField?: WorkflowsTimeField;
  query?: string;
}
type WorkflowsQuery = BaseQuery<'Workflows', WorkflowsOptions>
//#endregion

//#region Deployments Query
export type DeploymentsOptions = Options & {
  sha?: string;
  gitRef?: string;
  task?: string;
  environment?: string;
}
type DeploymentsQuery = BaseQuery<'Deployments', DeploymentsOptions>
//#endregion

export type GitHubQuery =
  Code_ScanningQuery |
  CommitsQuery |
  Commit_FilesQuery |
  IssuesQuery |
  ContributorsQuery |
  TagsQuery |
  ReleasesQuery |
  Pull_RequestsQuery |
  Pull_Request_ReviewsQuery |
  Pull_Request_FilesQuery |
  LabelsQuery |
  RepositoriesQuery |
  OrganizationsQuery |
  GraphQLQuery |
  MilestonesQuery |
  PackagesQuery |
  VulnerabilitiesQuery |
  ProjectsQuery |
  ProjectItemsQuery |
  StargazersQuery |
  Workflow_RunsQuery |
  Workflow_UsageQuery |
  WorkflowsQuery |
  DeploymentsQuery

export type GitHubVariableQuery = { key?: string; field?: string; } & GitHubQuery

export type GitHubAnnotationQuery = { timeField?: string; } & GitHubVariableQuery
